ansible_sdk/_util/metrics.py

```diff
@@ -0,0 +1,207 @@
+# Copyright: Ansible Project
+# Apache License 2.0 (see LICENSE or https://www.apache.org/licenses/LICENSE-2.0)
+
+import os
+import tarfile
+
+from datetime import datetime
+
+import ansible_sdk._util.dataclass_compat as dataclasses
+from ansible_sdk.model.job_status import AnsibleJobStatus
+from ansible_sdk._aiocompat.csv_async import CSVAsync
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class AnsibleJobStats:
+    job_id: str
+    job_type: str
+    started: str
+    finished: str
+    job_state: str
+    hosts_ok: str
+    hosts_changed: str
+    hosts_skipped: str
+    hosts_failed: str
+    hosts_unreachable: str
+    task_count: str
+    task_duration: str
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class AnsibleModuleStats:
+    job_id: str
+    module_fqcn: str
+    role_fqcn: str
+    task_count: int
+    task_duration: str
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class AnsibleCollectionStats:
+    job_id: str
+    collection_fqcn: str
+    collection_version: str
+    task_count: int
+    task_duration: str
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class AnsibleRoleStats:
+    job_id: str
+    role_fqcn: str
+    collection_version: str
+    task_count: int
+    task_duration: str
+
+
+@dataclasses.dataclass(frozen=False, kw_only=True)
+class MetricsData:
+    started: datetime = dataclasses.field(default_factory=datetime.now)
+    task_counter: int = 0
+    task_data: dict = dataclasses.field(default_factory=dict)
+    role_data: dict = dataclasses.field(default_factory=dict)
+    collection_data: dict = dataclasses.field(default_factory=dict)
+
+
+class MetricsCalc:
+    def __init__(self):
+        pass
+
+    async def collect_metrics(self, status_obj: AnsibleJobStatus):
+        metrics_data = MetricsData()
+        job_csv_filename = os.path.join(
+            status_obj._job_def.metrics_output_path, "jobs.csv"
+        )
+        module_csv_filename = os.path.join(
+            status_obj._job_def.metrics_output_path, "modules.csv",
+        )
+        collection_csv_filename = os.path.join(
+            status_obj._job_def.metrics_output_path, "collections.csv"
+        )
+        roles_csv_filename = os.path.join(
+            status_obj._job_def.metrics_output_path, "roles.csv"
+        )
+
+        job_headers = [x.name for x in dataclasses.fields(AnsibleJobStats)]
+        module_headers = [x.name for x in dataclasses.fields(AnsibleModuleStats)]
+        collection_headers = [x.name for x in dataclasses.fields(AnsibleCollectionStats)]
+        role_headers = [x.name for x in dataclasses.fields(AnsibleRoleStats)]
+
+        with open(job_csv_filename, "w") as job_csv_fh, \
```

Comment:
Wrapping the group of context managers in `()` is preferable to using line-continuation